### DISCUSSION TITLE
Lora directory help

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#11499] [Libraries & Drivers]
I'm using TTGO esp32  with onboard lora  that I setup with
```
from machine import Pin,SoftSPI,I2C,SoftI2C; from sx127x import SX127x
cnfg={'miso':19, 'mosi':27, 'ss':18, 'sck':5, 'dio_0':26, 'reset':14, 'led':25}
loracnfg={'frequency':9205E5, 'tx_power_level':14, 'signal_bandwidth':125E3, 'spreading_factor': 12, 'coding_rate':5, 'preamble_length':8, 'implicit_header':0, 'sync_word':0x12, 'enable_CRC':0, 'invert_IQ':0}
spi=SoftSPI(baudrate=10000000, polarity=0, phase=0, bits=8, firstbit=SoftSPI.MSB, sck=Pin(cnfg['sck'],Pin.OUT,Pin.PULL_DOWN), mosi=Pin(cnfg['mosi'],Pin.OUT,Pin.PULL_UP), miso=Pin(cnfg['miso'],Pin.IN,Pin.PULL_UP))
lora=SX127x(spi, pins=cnfg, parameters=loracnfg)
```
so far so good. But when I go looking for some commands that might be useful sometimes I get
```
>>> dir('lora')
['__class__', 'count', 'endswith', 'find', 'format', 'index', 'isalpha', 'isdigit', 'islower', 'isspace', 'isupper', 'join', 'lower', 'lstrip', 'replace', 'rfind', 'rindex', 'rsplit', 'rstrip', 'split', 'startswith', 'strip', 'upper', 'center', 'encode', 'partition', 'rpartition', 'splitlines']
```
but other times I get
```
>>> dir(lora)
['__class__', '__init__', '__module__', '__qualname__', 'write', '__dict__', 'sleep', 'default_parameters', 'enable_CRC', 'invert_IQ', '_spi', '_pins', '_parameters', '_lock', '_pin_rx_done', '_pin_ss', '_led_status', 'read_register', 'set_frequency', 'set_signal_bandwidth', 'write_register', 'set_tx_power', '_implicit_header_mode', 'implicit_header_mode', 'set_spreading_factor', 'set_coding_rate', 'set_preamble_length', 'set_sync_word', 'standby', 'begin_packet', 'end_packet', 'collect_garbage', 'set_lock', 'println', 'get_irq_flags', 'packet_rssi', '_frequency', 'packet_snr', '_tx_power_level', 'set_channel', 'dump_registers', 'receive', 'on_receive', 'handle_on_receive', 'read_payload', 'received_packet', 'transfer', 'blink_led']
```
Of the two the first one seems dodgy
```
>>> lora.count()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'SX127x' object has no attribute 'count'
```
I can't figure out why the 2 different responses as there seems to be only one sx127x
```
>>> import os; os.listdir(); os.listdir('lib')
['boot.py', 'lib']
['ssd1306.py', 'sx127x.py']
```

**Comment by jimmo** [comment]
> `>>> dir('lora')`

@kjm1102  This is doing a dir on the literal string `"lora"`. You mean `dir(lora)` (as you do in the second example).    (Maybe you're confused by `help("modules")` which is a special case for listing the installed modules).

**Reply by kjm1102** [reply]
Tnx Jim. I'm trying to figure out if the sx1272 can wakeup the esp32 from deepsleep after a lora packet has been received but after reading https://github.com/lemariva/uPyLoRaWAN/issues/16 twice it's still not clear to me if that is possible?

**Reply by davefes** [reply]
Interesting question.  Reading through RFM96W-V2.0.pdf there is a CAD mode see page 32.  On page 42 they mention that DIO1 can be configured to generate a CadDone interrupt.  I get the impression that the RFM96W can generate this without uP support, but it doesn't specifically state this.

I have just finished my first LoRa project using https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266
Maybe, worthwhile looking through his sx127x.py.

Good luck!

**Reply by kjm1102** [reply]
I don't understand cad Dave. https://lora-developers.semtech.com/documentation/tech-papers-and-guides/channel-activity-detection-ensuring-your-lora-packets-are-sent/how-to-ensure-your-lora-packets-are-sent-properly/ raves on about it being a way to avoid  'power-hungry continuous receiving' mode but if I put the TTGO into a while 1: receive loop waiting for a lora.packet_receive event the current does not change from when the board is just idling, about 34mA.

The HPD13A V1.1 module in the TTGO is an sx1276 so it should be able to detect the full packet without host MCU involvement, I just need to figure out how!

**Reply by davefes** [reply]
> This is where CAD shines, as a device will detect ongoing transmissions before sending its packet.

Sounds like a listen before transmit feature.  I played with RM22B chips years ago, but a LoRa beginner.

**Reply by davefes** [reply]
https://www.thethingsnetwork.org/forum/t/channel-activity-detection-cad-implementation-of-lora/13373?page=2

**Reply by kjm1102** [reply]
You don't happen to know a document that explains the difference between lora.on_receive & lora.handle_on_receive do you?

**Reply by davefes** [reply]
Provide a link to the code you are using.  I don't see lora.handle_on_receive in Sandeep's code.  In Wei's code onReceive() looks like it is used to generate an interrupt ... when the receive process is finished?  I think we need to generate an interrupt when you have got a preamble, to wake-up the ESP32 and proceed.

**Reply by kjm1102** [reply]
I went for the sx127x library at https://github.com/lemariva/uPyLoRaWAN Dave.  I Use it because the same guy had done a pretty good job with the esp32 camera board lib. I'll put all my other posts on https://github.com/orgs/micropython/discussions/11515 so our lora chat is all in one place.
